vhdx/src/vhdx_header.rs

```diff
@@ -0,0 +1,524 @@
+    // Checksum does not match
+    InvalidChecksum(String),
+    // Entry count does not match spec
+    InvalidEntryCount,
+    // Header signature does not match spec
+    InvalidHeaderSign,
+    // Region signature does not match spec
+    InvalidRegionSign,
+    // Not a valid UUID
+    InvalidUuid,
+    // Not a VHDx sign, i.e. not a VHDx file
+    InvalidVHDXSign,
+    // None of the two headers are valid
+    NoValidHeader,
+    // Failed reading checksum
+    ReadChecksum,
+    // Failed reading the File Type Identifier
+    ReadFileTypeIndetifier(io::Error),
+    // Failed reading the Header
+    ReadHeader(io::Error),
+    // Failed reading a Region Table Entry
+    ReadRegionTableEntry(io::Error),
+    // Failed reading the Region Table Header
+    ReadRegionTableHeader(io::Error),
+    // Failed gathering all region entries
+    RegionEntryCollectionFailed,
+    // Found overlap between region entries
+    RegionOverlap,
+    // This region is supposed to be zero
+    ReservedIsNonZero,
+    // This implementation does not recognize this entry
+    UnrecognizedRegionEntry,
+    // Failed writing Header to the file
+    WriteHeader(io::Error),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl Display for Error {
+    #[remain::check]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        #[sorted]
+        match self {
+            CalculateChecksum => write!(f, "Failed to calculate checksum"),
+            DuplicateBATEntry => write!(f, "BAT entry is not unique"),
+            DuplicateMDREntry => write!(f, "Metadata region entry is not unique"),
+            InvalidChecksum(section) => write!(f, "{} checksum doesn't match", section),
+            InvalidEntryCount => write!(f, "invalid entry cound"),
+            InvalidHeaderSign => write!(f, "not a valid VHDx header"),
+            InvalidRegionSign => write!(f, "not a valid VHDx region"),
+            InvalidUuid => write!(f, "couldn't parse Uuid for region entry"),
+            InvalidVHDXSign => write!(f, "not a VHDx file"),
+            NoValidHeader => write!(f, "No valid header found"),
+            ReadChecksum => write!(f, "cannot read checksum"),
+            ReadFileTypeIndetifier(e) => write!(f, "failed to read File Type Identifier: {}", e),
+            ReadHeader(e) => write!(f, "failed to read headers: {}", e),
+            ReadRegionTableEntry(e) => write!(f, "failed to read region table entry: {}", e),
+            ReadRegionTableHeader(e) => write!(f, "failed to read region table header: {}", e),
+            RegionEntryCollectionFailed => write!(f, "failed to read region entries"),
+            RegionOverlap => write!(f, "overlapping regions found"),
+            ReservedIsNonZero => write!(f, "reserved region has non-zero value"),
+            UnrecognizedRegionEntry => write!(f, "we do not recongize this entry"),
+            WriteHeader(e) => write!(f, "failed to write header: {}", e),
```

Comment:
Make sure you're consistent across error messages, including using a capital letter at the beginning.